### PR TITLE
Changes to the dbapi and the models for the Paths

### PR DESCRIPTION
**Pull request by sayanchowdhury**

### CODE REVIEW

File: tahrir_api/dbapi.py [hunk 1]
```diff
@@ -49,6 +50,248 @@ def __init__(self, dburi=None, session=None, autocommit=True,
...
+        """
+        Return the team with the given ID
+
+        :type team_id: str
+        :param team_id: The ID of the team to return
+        """
+
+        if self.team_exists(team_id):
+            return self.session.query(Team).filter(
+                func.lower(Team.id) == func.lower(team_id)).one()
```
Comment (pypingou) [inline tahrir_api/dbapi.py]: Thought about using ``.first()`` to do one SQL query instead of two?

File: tahrir_api/dbapi.py [hunk 2]
```diff
@@ -49,6 +50,248 @@ def __init__(self, dburi=None, session=None, autocommit=True,
...
+        :param name: position of the perk in the series
+
+        :type badge_id: str
+        :param badge_id: Badge ID for the Perk
+
+        :type series_id: str
+        :param series_id: ID of the Series
+        """
+        if self.perk_exists_for_badge_series(badge_id, series_id):
+            perk = self.get_perk_from_badge_series(badge_id, series_id).first()
```
Comment (pypingou) [inline tahrir_api/dbapi.py]: You're doing two queries here while you could be doing just one by using the ``.first()``, so something like:
~~~~
perk = self.get_perk_from_badge_series(badge_id, series_id).first()
if not perk:
     perk = Perk(...)
~~~~

File: tahrir_api/model.py [hunk 3]
```diff
@@ -58,21 +59,22 @@ def __json__(self):
...
     name = Column(Unicode(128), nullable=False, unique=True)
     image = Column(Unicode(128), nullable=False)
     stl = Column(Unicode(128))
     description = Column(Unicode(128), nullable=False)
     criteria = Column(Unicode(128), nullable=False)
+    issuer_id = Column(Integer, ForeignKey('issuers.id'), nullable=False)
+    perk = relationship("Perk", backref="badge")
     authorizations = relationship("Authorization", backref="badge")
     assertions = relationship("Assertion", backref="badge")
-    issuer_id = Column(Integer, ForeignKey('issuers.id'), nullable=False)
```
Comment (pypingou) [inline tahrir_api/model.py]: The changes here need an alembic migration script

File: tahrir_api/dbapi.py (hunk 2 above)
Comment (sayanchowdhury) [inline tahrir_api/dbapi.py]: Rightly said.

**Comment by sayanchowdhury** [comment]
@pypingou can you review this once. pushed the migration scripts too.

File: alembic/versions/e7040e76728_create_new_tables_for_quest.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import sqlalchemy as sa
+
+
+def upgrade():
+
+    op.create_table(
+        'team',
+        sa.Column('id', sa.Unicode(length=128), nullable=False),
+        sa.Column('name', sa.Unicode(length=128), nullable=False),
+        sa.Column('created_on', sa.DateTime(), nullable=False),
```
Comment (pypingou) [inline alembic/versions/e7040e76728_create_new_tables_for_quest.py]: Shouldn't it have a default?

File: alembic/versions/e7040e76728_create_new_tables_for_quest.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        sa.PrimaryKeyConstraint('id'),
+        sa.UniqueConstraint('name')
+    )
+    op.create_table(
+        'series',
+        sa.Column('id', sa.Unicode(length=128), nullable=False),
+        sa.Column('name', sa.Unicode(length=128), nullable=False),
+        sa.Column('description', sa.Unicode(length=128), nullable=False),
+        sa.Column('created_on', sa.DateTime(), nullable=False),
+        sa.Column('last_updated', sa.DateTime(), nullable=False),
```
Comment (pypingou) [inline alembic/versions/e7040e76728_create_new_tables_for_quest.py]: Shouldn't it have `onupdate` instructions?

**Comment by pypingou** [comment]
Reviewed the last 4 commits, only got comments about the migration script :)

**Comment by sayanchowdhury** [comment]
@pypingou Can you give a quick look at the changes?

File: alembic/versions/e7040e76728_create_new_tables_for_quest.py [hunk 6]
```diff
@@ -10,27 +10,37 @@
...
+def generate_default_id(context):
+    return context.current_parameters['name'].lower().replace(' ', '-')
 
 def upgrade():
 
     op.create_table(
         'team',
-        sa.Column('id', sa.Unicode(length=128), nullable=False),
+        sa.Column('id', sa.Unicode(length=128), nullable=False,
+                  default=generate_default_id),
```
Comment (pypingou) [inline alembic/versions/e7040e76728_create_new_tables_for_quest.py]: Not entirely sure to follow what this does, could you expand?

Comment (sayanchowdhury) [inline alembic/versions/e7040e76728_create_new_tables_for_quest.py]: I have added this [method](http://docs.sqlalchemy.org/en/latest/core/defaults.html#context-sensitive-default-functions) in the migration file itself. It slugifies the input based on the name.

Comment (pypingou) [inline alembic/versions/e7040e76728_create_new_tables_for_quest.py]: Oh I see, neat! :)

**Comment by pypingou** [comment]
Looks alright to me :)

**Pull request merged**